book/visual-effects.md

```diff
@@ -5,312 +5,396 @@ prev: security
-integration.
+Next, we need create an SDL window, instead of a Tkinter window,
+inside the Browser, and set up Skia to draw to it. Here's the SDL
+incantation to create a window:
 
-::: {.further}
-While this book is about browsers, and not how to implement high-quality
-raster libraries, that topic is very interesting in its own right.
-In addition, it is very important these days for browsers to work smoothly with
-the advanced GPUs in today's devices, and often browsers are pushing the
-envelope of graphics technology. So in practice browser teams include experts
-in these areas: Skia for Chromium and [Core Graphics][core-graphics] for Webkit,
-for example. In both cases these libraries are used outside of the
-browser---Core Graphics is used for iOS and macOS apps, and Skia for Android.
-:::
+``` {.python}
+class Browser:
+    def __init__(self):
+        self.sdl_window = sdl2.SDL_CreateWindow(b"Browser",
+            sdl2.SDL_WINDOWPOS_CENTERED, sdl2.SDL_WINDOWPOS_CENTERED,
+            WIDTH, HEIGHT, sdl2.SDL_WINDOW_SHOWN)
+```
 
-[core-graphics]: https://developer.apple.com/documentation/coregraphics
+To set up Skia to draw to this window, we need to create two
+"surfaces":[^surface]
 
-To install Skia, you'll need to install the
-[`skia-python`](https://github.com/kyamagu/skia-python)
-library (via `pip3 install skia-python`); as explained on the linked site, you
-might need to install additional dependencies. Instructions
-[here](https://pypi.org/project/PySDL2/) explain how to install SDL on your
-computer (short version: `pip3 install PySDL2` and `pip3 install pysdl2-dll`).
+[^surface]: In Skia and SDL, a *surface* is a representation of a
+graphics buffer into which you can draw "pixels" (bits representing
+colors). A surface may or may not be bound to the actual pixels on the
+screen via a window, and there can be many surfaces. A *canvas* is an
+API interface that allows you to draw into a surface with higher-level
+commands such as for rectangles or text. Our browser uses separate
+Skia and SDL surface for simplicity, but in a highly optimized
+browser, minimizing the number of surfaces is important for good
+performance.
 
-Once installed, remove `tkinter` from your Python imports and replace them with:
 
 ``` {.python}
-import ctypes
-import sdl2
-import skia
+class Browser:
+    def __init__(self):
+        self.window_surface = sdl2.SDL_GetWindowSurface(
+            self.sdl_window)
+        self.skia_surface = skia.Surface(WIDTH, HEIGHT)
 ```
 
-The additional `ctypes` module is for interfacing between Python types and C
-types.
+Typically, we'lll draw to the Skia surface, and then once we're done
+with it we'll copy it to the SDL surface to display on the screen.
+This looks a little hairy but really it's just copying pixels from one
+place to another:
 
-The main loop of the browser first needs some boilerplate to get SDL started:
+``` {.python}
+class Browser:
+    def commit_canvas(self):
```

Comment:
```suggestion
    def draw_to_screen(self):
```

"draw" is what Chromium calls the process of taking the rastered pixels and putting them onto the screen
(usually with non-trivial GL programs to do it). But the Skia canvas API has various methods called "draw",
so I suggest "draw_to_screen" as a way to make it clear.